File: src/main/java/me/alegian/thavma/impl/client/renderer/entity/VisER.java
package me.alegian.thavma.impl.client.renderer.entity;

import com.mojang.blaze3d.vertex.PoseStack;
import com.mojang.math.Axis;
import me.alegian.thavma.impl.client.T7PoseStack;
import me.alegian.thavma.impl.client.renderer.VisRenderer;
import me.alegian.thavma.impl.common.entity.VisEntity;
import net.minecraft.client.Minecraft;
import net.minecraft.client.player.LocalPlayer;
import net.minecraft.client.renderer.MultiBufferSource;
import net.minecraft.client.renderer.culling.Frustum;
import net.minecraft.client.renderer.entity.EntityRenderer;
import net.minecraft.client.renderer.entity.EntityRendererProvider;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.entity.HumanoidArm;
import net.minecraft.world.inventory.InventoryMenu;
import net.minecraft.world.phys.Vec3;
import net.neoforged.api.distmarker.Dist;
import net.neoforged.api.distmarker.OnlyIn;

@OnlyIn(Dist.CLIENT)
public class VisER extends EntityRenderer<VisEntity> {

  public VisER(EntityRendererProvider.Context pContext) {
    super(pContext);
  }

  @Override
  public void render(VisEntity visEntity, float pEntityYaw, float pPartialTick, PoseStack pPoseStack, MultiBufferSource pBufferSource, int pPackedLight) {
    if (visEntity.getPlayer() == null) return;

    T7PoseStack t7pose = new T7PoseStack(pPoseStack);
    t7pose.push();
    t7pose.translateNegative(visEntity.position()); // we are inside an entity renderer

    preparePlayerHandPose(pPartialTick, (LocalPlayer) visEntity.getPlayer(), t7pose);

    VisRenderer.render(visEntity.position(), t7pose, pBufferSource, visEntity.tickCount + pPartialTick);

    t7pose.pop();
  }

  /**
   * Translate the Pose to the Player's hand. This is done approximately, and does not follow
   * the PlayerModel's animations, because using these animation poses introduces rotations and
   * reflections that the VisRenderer cannot handle.
   */
  private static void preparePlayerHandPose(float pPartialTick, LocalPlayer player, T7PoseStack t7pose) {
    Vec3 playerPos = player.getPosition(pPartialTick);
    t7pose.translate(playerPos);

    var arm = player.getMainArm();

    // for first person, if it is the client player, we follow the camera
    if (player == Minecraft.getInstance().player && Minecraft.getInstance().options.getCameraType().isFirstPerson()) {
      var angle = Math.PI / 2 - player.getViewYRot(pPartialTick) / 360F * 2 * Math.PI;
      var translation = player.getViewVector(pPartialTick).normalize().scale(.1f);
      t7pose.translate(0, player.getEyeHeight() + 0.01, 0);
      t7pose.translate(translation);
      t7pose.rotate(Axis.YP, angle);
      t7pose.translate(0, 0, arm == HumanoidArm.RIGHT ? -.06f : .06f);
      t7pose.rotate(Axis.YP, -angle);
    } else { // for third person, we follow body rotation
      var angle = Math.PI / 2 - player.getPreciseBodyRotation(pPartialTick) / 360F * 2 * Math.PI;
      t7pose.rotate(Axis.YP, angle);
      t7pose.translate(-1f, player.getEyeHeight() - .56f, arm == HumanoidArm.RIGHT ? -.4f : .4f);
      t7pose.rotate(Axis.YP, -angle);
    }
  }

  @Override
  public ResourceLocation getTextureLocation(VisEntity pEntity) {
    return InventoryMenu.BLOCK_ATLAS;
  }

  /**
   * The Vis Entity does not have a strict bounding box,
   * so we never cull it to avoid rendering bugs at the edge
   * of the screen.
   */
  @Override
  public boolean shouldRender(VisEntity pLivingEntity, Frustum pCamera, double pCamX, double pCamY, double pCamZ) {
    return true;
  }
}

File: src/main/java/me/alegian/thavma/impl/client/renderer/entity/VisER.kt
package me.alegian.thavma.impl.client.renderer.entity

import com.mojang.blaze3d.vertex.PoseStack
import me.alegian.thavma.impl.client.renderer.level.renderEssentia
import me.alegian.thavma.impl.client.util.translate
import me.alegian.thavma.impl.common.entity.VisEntity
import net.minecraft.client.Minecraft
import net.minecraft.client.renderer.MultiBufferSource
import net.minecraft.client.renderer.culling.Frustum
import net.minecraft.client.renderer.entity.EntityRenderer
import net.minecraft.client.renderer.entity.EntityRendererProvider
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.entity.HumanoidArm
import net.minecraft.world.entity.player.Player
import net.minecraft.world.inventory.InventoryMenu
import net.minecraft.world.phys.Vec3
import net.neoforged.api.distmarker.Dist
import net.neoforged.api.distmarker.OnlyIn
import thedarkcolour.kotlinforforge.neoforge.forge.vectorutil.v3d.plus
import thedarkcolour.kotlinforforge.neoforge.forge.vectorutil.v3d.unaryMinus

@OnlyIn(Dist.CLIENT)
class VisER(pContext: EntityRendererProvider.Context) : EntityRenderer<VisEntity>(pContext) {
    override fun render(visEntity: VisEntity, pEntityYaw: Float, pPartialTick: Float, poseStack: PoseStack, pBufferSource: MultiBufferSource, pPackedLight: Int) {
        val player = visEntity.player ?: return

        poseStack.pushPose()
        poseStack.translate(-visEntity.position()) // we are inside an entity renderer
        val playerHandPos = preparePlayerHandPosition(pPartialTick, player)

        renderEssentia(visEntity.position(), playerHandPos, poseStack, pBufferSource, visEntity.tickCount + pPartialTick)
        poseStack.popPose()
    }

    override fun getTextureLocation(pEntity: VisEntity): ResourceLocation {
        return InventoryMenu.BLOCK_ATLAS
    }

    /**
     * The Vis Entity does not have a strict bounding box,
     * so we never cull it to avoid rendering bugs at the edge
     * of the screen.
     */
    override fun shouldRender(pLivingEntity: VisEntity, pCamera: Frustum, pCamX: Double, pCamY: Double, pCamZ: Double): Boolean {
        return true
    }
}

/**
 * Prepare a Vec3 position of the Player's hand. This is only an approximation,
 * and it does not follow the PlayerModel's animations, because using these
 * animation poses is complicated.
 */
private fun preparePlayerHandPosition(pPartialTick: Float, player: Player): Vec3 {
    var position = player.getPosition(pPartialTick)

    val arm = player.mainArm

    // for first person, if it is the client player, we follow the camera
    if (player === Minecraft.getInstance().player && Minecraft.getInstance().options.cameraType.isFirstPerson) {
        val angle = Math.PI / 2 - player.getViewYRot(pPartialTick) / 360f * 2 * Math.PI
        val translation = player.getViewVector(pPartialTick).normalize().scale(.1)
        position += Vec3(0.0, player.eyeHeight + 0.01, 0.0)
        position += translation
        val horizontalOffset = Vec3(0.0, 0.0, (if (arm == HumanoidArm.RIGHT) -.06f else .06f).toDouble())
        position += horizontalOffset.yRot(angle.toFloat())
    } else { // for third person, we follow body rotation
        val angle = Math.PI / 2 - player.getPreciseBodyRotation(pPartialTick) / 360f * 2 * Math.PI
        val offset = Vec3(-1.0, (player.eyeHeight - .56f).toDouble(), (if (arm == HumanoidArm.RIGHT) -.4f else .4f).toDouble())
        position += offset.yRot(angle.toFloat())
    }
    return position
}
